libobs/obs-audio-controls.c

```diff
@@ -256,48 +259,254 @@ static void volmeter_source_destroyed(void *vptr, calldata_t *calldata)
+
+		SHIFT_RIGHT_2PS(new_work, work);
+		VECTOR_MATRIX_CROSS_PS(intrp_samples, work, m3, m1, p1, p3);
+		peak = _mm_max_ps(peak, abs_ps(intrp_samples));
+
+		SHIFT_RIGHT_2PS(new_work, work);
+		VECTOR_MATRIX_CROSS_PS(intrp_samples, work, m3, m1, p1, p3);
+		peak = _mm_max_ps(peak, abs_ps(intrp_samples));
+	}
+
+	return hmax_ps(peak);
+}
+
+/* points contain the first four samples to calculate the sinc interpolation
+ * over. They will have come from a previous iteration.
+ */
+static float get_sample_peak(__m128 previous_samples, const float *samples,
+	size_t nr_samples)
+{
+	__m128 peak = previous_samples;
+	for (size_t i = 0; (i + 3) < nr_samples; i += 4) {
+		__m128 new_work = _mm_load_ps(&samples[i]);
+		peak = _mm_max_ps(peak, abs_ps(new_work));
+	}
+	return hmax_ps(peak);
+}
+
+static void volmeter_process_peak_last_samples(obs_volmeter_t *volmeter,
+		int channel_nr, float *samples, size_t nr_samples)
+{
+	// Take the last 4 samples that need to be used for the next peak
+	// calculation. If there are less than 4 samples in total the new
+	// samples shift out the old sampless.
+
+	switch (nr_samples) {
+	case 0:
+		break;
+	case 1:
+		volmeter->prev_samples[channel_nr][0] =
+				volmeter->prev_samples[channel_nr][1];
+		volmeter->prev_samples[channel_nr][1] =
+				volmeter->prev_samples[channel_nr][2]; 
+		volmeter->prev_samples[channel_nr][2] =
+				volmeter->prev_samples[channel_nr][3];
+		volmeter->prev_samples[channel_nr][3] = samples[nr_samples-1];
+		break;
+	case 2:
+		volmeter->prev_samples[channel_nr][0] =
+				volmeter->prev_samples[channel_nr][2];
+		volmeter->prev_samples[channel_nr][1] =
+				volmeter->prev_samples[channel_nr][3]; 
+		volmeter->prev_samples[channel_nr][2] = samples[nr_samples-2];
+		volmeter->prev_samples[channel_nr][3] = samples[nr_samples-1];
+		break;
+	case 3:
+		volmeter->prev_samples[channel_nr][0] =
+				volmeter->prev_samples[channel_nr][3];
+		volmeter->prev_samples[channel_nr][1] = samples[nr_samples-3];
+		volmeter->prev_samples[channel_nr][2] = samples[nr_samples-2];
+		volmeter->prev_samples[channel_nr][3] = samples[nr_samples-1];
+		break;
+	default:
+		volmeter->prev_samples[channel_nr][0] = samples[nr_samples-4];
+		volmeter->prev_samples[channel_nr][1] = samples[nr_samples-3];
+		volmeter->prev_samples[channel_nr][2] = samples[nr_samples-2];
+		volmeter->prev_samples[channel_nr][3] = samples[nr_samples-1];
+	}
+}
+
+static void volmeter_process_peak(obs_volmeter_t *volmeter,
+		const struct audio_data *data, int nr_channels)
 {
 	int nr_samples = data->frames;
 	int channel_nr = 0;
-
-	for (size_t plane_nr = 0; plane_nr < MAX_AV_PLANES; plane_nr++) {
+	for (int plane_nr = 0; channel_nr < nr_channels; plane_nr++) {
 		float *samples = (float *)data->data[plane_nr];
 		if (!samples) {
-			// This plane does not contain data.
 			continue;
 		}
+		if (((off_t)samples & 0xf) > 0) {
+			printf("Audio plane %i is not aligned %p skipping peak volume measurement.\n", plane_nr, samples);
```

Comment:
Typically a string like this can be split up in to multiple lines.  Currently this line goes over the column limit.